// n8n-nodes-sii-chile/dist/credentials/SiiChileApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SiiChileApi = void 0;
class SiiChileApi {
    constructor() {
        this.name = 'siiChileApi';
        this.displayName = 'SII Chile / Electronic Invoicing';
        this.documentationUrl = 'https://www.openfactura.cl/factura-electronica/api/';
        this.properties = [
            {
                displayName: 'Invoicing Provider',
                name: 'provider',
                type: 'options',
                options: [
                    {
                        name: 'OpenFactura (Haulmer)',
                        value: 'openfactura',
                    },
                    {
                        name: 'SimpleAPI',
                        value: 'simpleapi',
                    },
                    {
                        name: 'Local Functions Only (no issuing)',
                        value: 'none',
                    },
                ],
                default: 'none',
                description: 'Service to issue DTEs to SII',
            },
            {
                displayName: 'OpenFactura API Key',
                name: 'openfacturaApiKey',
                type: 'string',
                typeOptions: {
                    password: true,
                },
                default: '',
                displayOptions: {
                    show: {
                        provider: ['openfactura'],
                    },
                },
                description: 'Get it at panel.openfactura.cl',
            },
            {
                displayName: 'OpenFactura Environment',
                name: 'openfacturaEnvironment',
                type: 'options',
                options: [
                    {
                        name: 'Production',
                        value: 'production',
                    },
                    {
                        name: 'Certification (Testing)',
                        value: 'certification',
                    },
                ],
                default: 'certification',
                displayOptions: {
                    show: {
                        provider: ['openfactura'],
                    },
                },
            },
            {
                displayName: 'SimpleAPI API Key',
                name: 'simpleapiApiKey',
                type: 'string',
                typeOptions: {
                    password: true,
                },
                default: '',
                displayOptions: {
                    show: {
                        provider: ['simpleapi'],
                    },
                },
                description: 'Get it at simpleapi.cl (has free plan)',
            },
            {
                displayName: 'Issuer RUT',
                name: 'rutEmisor',
                type: 'string',
                default: '',
                placeholder: '76123456-7',
                description: 'Your issuing company RUT',
                displayOptions: {
                    hide: {
                        provider: ['none'],
                    },
                },
            },
        ];
        this.test = {
            request: {
                baseURL: 'https://mindicador.cl',
                url: '/api',
                method: 'GET',
            },
        };
    }
}
exports.SiiChileApi = SiiChileApi;
